modelhub/modelhub/series/series_objectiv.py

```diff
@@ -63,6 +70,68 @@ def _bigquery_get_from_context_with_type_series(self, type: str, key: str, dtype
         value_series = ctx_series.json.get_value(key=key, as_str=as_str)
         return value_series.copy_override_dtype(dtype)
 
+    def filter_keys_of_dicts(self, keys: List[str]) -> 'TSeriesJson':
+        """
+        Return a new Series, that consists of the same top level array, but with all the sub-dictionaries
+        having only their original fields if that field is listed in the keys parameter.
+        """
+        engine = self._series_object.engine
+        if is_postgres(engine):
+            return self._pg_filter_keys_of_dicts(keys)
+        if is_bigquery(engine):
+            return self._bigquery_filter_keys_of_dicts(keys)
+        raise DatabaseNotSupportedException(engine)
+
+    def _pg_filter_keys_of_dicts(self, keys: List[str]) -> 'TSeriesJson':
+        jsonb_build_object_str = [f"{quote_string(self._series_object.engine, key)}" for key in keys]
+        expression_str = f'''(
+            select jsonb_agg((
+                select json_object_agg(items.key, items.value)
+                from jsonb_each(objects.value) as items
+                where items.key in ({", ".join(jsonb_build_object_str)})))
+            from jsonb_array_elements({{}}) as objects)
+        '''
+        expression = Expression.construct(
+            expression_str,
+            self._series_object
+        )
+        return self._series_object.copy_override(expression=expression)
+
+    def _bigquery_filter_keys_of_dicts(self, keys: List[str]) -> 'TSeriesJson':
+        json_object_str_expressions = []
+        # We unnest the json-array, and then for every json object in the array we build a copy of that
+        # json object, but with only the keys that are listed in the `keys` variable.
+        # Unfortunately BigQuery has no way (yet) to turn a struct into a json string, so we have to do raw
+        # string building to get json objects.
+        for i, key in enumerate(keys):
+            if '"' in key:
+                raise ValueError(f'key values containing double quotes are not supported. key: {key}')
+            key_expr = Expression.construct('{}', Expression.string_value(json.dumps(key)))
+            colon_expr = Expression.string_value(': ')
+            val_expr = Expression.construct('''JSON_QUERY(item, '$."{}"')''', Expression.raw(key))
+            comma_expr = Expression.string_value(', ')
+            json_object_str_expressions.append(key_expr)
+            json_object_str_expressions.append(colon_expr)
+            json_object_str_expressions.append(val_expr)
+            if i < (len(keys) - 1):
+                json_object_str_expressions.append(comma_expr)
+        json_object_key_values_expr = Expression.construct(
+            ', '.join('{}' for _ in json_object_str_expressions),
+            *json_object_str_expressions
+        )
```

Comment:
I felt that was a bit difficult to follow since we have a lot of expressions and we kinda need to know when to add a comma.  Maybe building **key: value** expressions first and later join them all?
Something like:
```suggestion
        engine = self._series_object.engine
        key_template_expr = f"{{}} || {quote_string(engine, ':')} "
        unnested_item_helper_series = self._series_object.copy_override(expression=Expression.construct('item'))

        # creates expressions in key: value format
        sub_items_expr = [
            join_expressions(
                [
                    Expression.construct(key_template_expr, Expression.string_value(json.dumps(key))),
                    unnested_item_helper_series.json.get_value(key).expression,
                ],
                join_str=" || ",
            )
            for key in keys
        ]

        # add comma separator between each key-value
        json_object_key_values_expr = join_expressions(sub_items_expr, join_str=f" || {quote_string(engine, ',')} || ")
```

Reply:
I get what you are saying, this is difficult to follow. However I found your version even harder to follow.

I have made a new attempt, used some of your improvements (e.g. `join_expressions()`), and added more clarifying comments, but it is essentially the same algorithm and gives the exact same output. What do you think..

This is the commit with the changes: https://github.com/objectiv/objectiv-analytics/pull/855/commits/88e748acea61e48e0ee501d5b302fc0ffc7531fc

Reply:
Yes it is, but still. Looks odd to check when to add a comma, as you could just use `join_str` parameter for separating each key-value expression by: `"|| {quote_string(',')} ||"`. I think that my main problem is just having too many individual  expressions that can be grouped into one, I guess that's why it was difficult for me to understand as I had to keep track of the position of each of them.  But its a subjective matter ig, I have no problem to merge it as it is as this is more of a provisional solution and it'll  be moved to bach (soonish, I hope).

Reply:
I think we might want to use `to_json_string`. Then we don't need to bother about the commas etc.

I didn't test it with expressions etc, but the query for `feature_stack` could then look like change from (only the part that creates the array, so not the last part of creating a string from the array):
```sql
ARRAY(select concat('{', """\"_type\"""", """: """, JSON_QUERY(item, '$."_type"'), """, """, """\"id\"""", """: """, JSON_QUERY(item, '$."id"'), '}') from unnest(json_query_array(`location_stack`, '$')) as item)
```
to:
```sql
ARRAY(select to_json_string(struct(JSON_EXTRACT_SCALAR(item, "$._type") as _type, JSON_EXTRACT_SCALAR(item, "$.id") as id)) from unnest(json_query_array(`location_stack`, '$')) as item)
```

Reply:
I think this is a very good suggestion, and much simpler.

Only concern I'd have is with the `JSON_EXTRACT_SCALAR`, it will only work if the fields are string, numeric, or boolean. If the context contains a nested json object it won't work. But looking at the schema we don't currently have that, and I guess we won't have that? @hendrik-obj 

I'll go ahead and implement your suggestion then

Reply:
follow-up issue: https://github.com/objectiv/objectiv-analytics/issues/909